qiskit/qobj/_qobj.py

```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""Models for QObj and its related components."""
+
+from types import SimpleNamespace
+
+from ._utils import QObjType, QObjValidationError
+
+# Current version of the QObj schema.
+QOBJ_VERSION = '0.0.1'
+
+
+class QObjItem(SimpleNamespace):
+    """Generic QObj structure.
+
+    Single item of a QObj structure, acting as a superclass of the rest of the
+    more specific elements.
+    """
+    REQUIRED_ARGS = ()
+
+    def as_dict(self):
+        """
+        Return a dictionary representation of the QObjItem, recursively
+        converting its public attributes.
+
+        Returns:
+            dict: a dictionary.
+        """
+        return {key: self._expand_item(value) for key, value
+                in self.__dict__.items() if not key.startswith('_')}
+
+    @classmethod
+    def _expand_item(cls, obj):
+        """
+        Return a valid representation of `obj` depending on its type.
+        """
+        if isinstance(obj, list):
+            return [cls._expand_item(item) for item in obj]
+        if isinstance(obj, QObjItem):
+            return obj.as_dict()
+        return obj
+
+    @classmethod
+    def from_dict(cls, obj):
+        """
+        Return a QObjItem from a dictionary recursively, checking for the
+        required attributes.
+
+        Returns:
+            QObjItem: a new QObjItem.
+
+        Raises:
+            QObjValidationError: if the dictionary does not contain the
+                required attributes for that class.
+        """
+        if not all(key in obj.keys() for key in cls.REQUIRED_ARGS):
+            raise QObjValidationError(
+                'The dict does not contain all required keys: missing "%s"' %
+                [key for key in cls.REQUIRED_ARGS if key not in obj.keys()])
+
+        return cls(**{key: cls._qobjectify_item(value)
+                      for key, value in obj.items()})
+
+    @classmethod
+    def _qobjectify_item(cls, obj):
+        """
+        Return a valid value for a QObjItem from a object.
+        """
+        if isinstance(obj, dict):
+            # TODO: should use the subclasses for finer control over the
+            # required arguments.
+            return QObjItem.from_dict(obj)
+        elif isinstance(obj, list):
+            return [cls._qobjectify_item(item) for item in obj]
+        return obj
+
+    def __reduce__(self):
+        """
+        Customize the reduction in order to allow serialization, as the QObjs
+        are automatically serialized due to the use of futures.
+        """
+        init_args = tuple(getattr(self, key) for key in self.REQUIRED_ARGS)
+        extra_args = {key: value for key, value in self.__dict__.items()
+                      if key not in self.REQUIRED_ARGS}
+        return self.__class__, init_args, extra_args
```

Comment:
There is probably a good reason, but why are you returning the required args first, then the extra args?

Reply:
Basically for making the serialization work as closely as possible to the `__init__` instantiation - [this method](https://docs.python.org/3/library/pickle.html#object.__reduce__) is a bit magic, but it is needed in order to have serialization working (which in turn, we need because the `Qobj`s are passed to the futures, and the default serialization fails).